src/_macosx.m

```diff
@@ -1000,74 +992,62 @@ -(void)zoom:(id)sender
-{   PyObject* canvas;
-    View* view;
-    PyObject* size;
-    NSRect rect;
-    int width, height;
+{   PyObject* canvas = NULL;
+    PyObject *mod = NULL, *base_class = NULL, *ret = NULL;
+    PyObject *subplot_tool = NULL, *subplot_figure = NULL, *subplot_canvas = NULL;
 
-    rect.origin.x = 100;
-    rect.origin.y = 350;
     PyGILState_STATE gstate = PyGILState_Ensure();
-    PyObject* master = PyObject_GetAttrString(toolbar, "canvas");
-    if (master==nil)
-    {
+    canvas = PyObject_GetAttrString(toolbar, "canvas");
+    if (!canvas) {
         PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
+        goto exit;
     }
-    canvas = PyObject_CallMethod(toolbar, "prepare_configure_subplots", "");
-    if(!canvas)
+
+    if (!((FigureCanvas*)canvas)->view) /* Something really weird going on */
     {
+        PyErr_SetString(PyExc_RuntimeError, "NSView* is NULL");
         PyErr_Print();
-        Py_DECREF(master);
-        PyGILState_Release(gstate);
-        return;
+        goto exit;
     }
 
-    view = ((FigureCanvas*)canvas)->view;
-    if (!view) /* Something really weird going on */
-    {
-        PyErr_SetString(PyExc_RuntimeError, "NSView* is NULL");
+    /* Prepare the subplot tool via the superclass. */
+    if (!(mod = PyImport_ImportModule("matplotlib.backend_bases"))
+        || !(base_class = PyObject_GetAttrString(mod, "NavigationToolbar2"))
+        || !(ret = PyObject_CallMethod(base_class, "configure_subplots",
+                                       "O", toolbar))) {
         PyErr_Print();
-        Py_DECREF(canvas);
-        Py_DECREF(master);
-        PyGILState_Release(gstate);
-        return;
+        goto exit;
     }
 
-    size = PyObject_CallMethod(canvas, "get_width_height", "");
-    Py_DECREF(canvas);
-    if(!size)
-    {
+    /* The subplot_tool attribute is set by the superclass method. */
+    if (!(subplot_tool = PyObject_GetAttrString(toolbar, "subplot_tool"))
+        || !(subplot_figure = PyObject_GetAttrString(subplot_tool, "figure"))
+        || !(subplot_canvas = PyObject_GetAttrString(subplot_figure, "canvas"))) {
         PyErr_Print();
-        Py_DECREF(master);
-        PyGILState_Release(gstate);
-        return;
+        goto exit;
     }
 
-    int ok = PyArg_ParseTuple(size, "ii", &width, &height);
-    Py_DECREF(size);
-    if (!ok)
+    if (!((FigureCanvas*)subplot_canvas)->view) /* Something really weird going on */
     {
+        PyErr_SetString(PyExc_RuntimeError, "NSView* is NULL");
         PyErr_Print();
-        Py_DECREF(master);
-        PyGILState_Release(gstate);
-        return;
+        goto exit;
     }
 
-    NSWindow* mw = [((FigureCanvas*)master)->view window];
-    Py_DECREF(master);
-    PyGILState_Release(gstate);
+    NSWindow* parent = [((FigureCanvas*)canvas)->view window];
+    NSWindow* child = [((FigureCanvas*)subplot_canvas)->view window];
+    [parent addChildWindow: child ordered: NSWindowAbove];
```

Comment:
I would remove the `addChildWindow` here. When it is attached to the parent the window will move around relative to it when dragged. This causes issues on dual monitors, where it is hidden from view when I drag the subplots window onto a new monitor. Removing this makes it behave as expected still I think, so I'm not sure there is a real benefit to attaching them.

Reply:
But does it close when you close the figure window? Even with multiple figures open?

Reply:
Ah, I see Qt doesn't handle that properly either.

Reply:
Yeah, this is a tricky one. I suggested the easy way out here ;) I think you could go down a rabbit hole that probably isn't super critical.

As it currently is, closing the figure window does close the subplot window as well, but the event-loop keeps running for some reason, so my Python process doesn't end.

Then I go and click on the still-running Python process to bring it back up and all of a sudden two windows appear again, with a phantom figure window that I can't do anything with and the subplots window that does work. I can then close the subplots window and it closes the entire session then.

Reply:
OK, I changed it to only set a dock window style, though I'm not sure it really had much effect.

I'll open a separate PR for the subplot window lifetime.

Reply:
I'm not sure you even need that?

I get some logs output saying it isn't supported. (It didn't fail, just logged)
```
python[68679:18018136] NSWindow does not support utility styleMask 0x10
```

Reply:
Oh, you're right; I missed that when testing. The docs don't seem to say that's a requirement, but okay.

Reply:
Looks good to me now. I'll open some new issues with the other things we found here.